rclc/src/rclc/executor.c

```diff
@@ -923,10 +974,28 @@ _rclc_execute(rclc_executor_handle_t * handle)
   if (invoke_callback) {
     switch (handle->type) {
       case SUBSCRIPTION:
-        if (handle->data_available) {
-          handle->callback(handle->data);
-        } else {
-          handle->callback(NULL);
+        switch (handle->callback_type) {
+          case CB_WITHOUT_REQUEST_ID:
```

Comment:
Should this be something like `CB_WITHOUT_CONTEXT`?

Reply:
Yes.

`CB_WITHOUT_REQUEST_ID` comes from when request_id was added to services in Nov2020, 
https://github.com/ros2/rclc/commit/4bc9c13ce03160c883c613fd99610838d928d1a5#diff-682b2c9ebab8a01f15a96d651ebc0e2aec2748f7abfafe288a48f9eb9fcc8786R48-R53
`CB_WITHOUT_REQUEST_ID` was treated like a default `callback_type` for the original api. It made sense when the request ID was the only special case, and only for services

I don't think it needs another callback-type, just a better name.

`handle.callback_type` is never checked without first checking `handle.type`, so I can also see an argument for eliminating `handle.callback_type` and extending `handle.type` to match 1:1 with all the types in the handle's function-pointer union: https://github.com/ros2/rclc/blob/c7e09de6c22f349ec456e7b966a3c907f516d894/rclc/include/rclc/executor_handle.h#L139-L148

Reply:
Yes, I agree. Initally, we only had services, subscriptions, clients. Now we have multiple types of callbacks for subscriptions, services and clients:
- with request id/without request id (for services/clients), 
- with context, without context (for subscriptions/clients/services). 

It would make sense to refactor handle->type and add e..g
SERVICE_WITH_REQ_ID
SERVICE_WITHOUT_REQ_ID
SERVICE_WITHOUT_CONTEXT
SERVICE_WITH_CONTEXT
SUBSCRIPTION_WITH_CONTEXT
SUBSCRIPTION_WITHOUT_CONTEXT
same for CLIENTS ...
(quite a number of types ...)

the request-id is currently not used in services with context. 
@norro @pablogs9 @BrettRD I propose to keep the structure like this in this pull request. Let's open a new issue for discussing how to simplify `callback-type` and `handle-type` variables. 

@norro. It looks like that the implementation of the client callback with context is not properly called: it only calls the client callback without context:
https://github.com/ros2/rclc/blob/0263722fb48be9bddadef3601c50397c8d71a1d8/rclc/src/rclc/executor.c#L976
You never use a client callback with context?

Reply:
Added an issue for that :https://github.com/ros2/rclc/issues/115